### DISCUSSION TITLE
Question about architecture

### DISCUSSION THREAD

**Discussion by halsten-dev** [fyne-io/fyne#4793] [Q&A] [answered]
Hi everyone, 

I'm currently learning Fyne and I find it nice but have some issue with the way to organize my code.

Of course I want to avoid the "all" in the main function. Separating my code into multiples files.

Actually, my question is pretty simple : How to pass window ref around ?

Explanation : 
For exemple, I have a button that need to change the content of the window.
To write nicer code (in my opinion), I want to have the function for my button as a separate function. So let's review some code : 

```go
func NewContent() *fyne.Container {
	btTest := widget.NewButton("Test", btTestClicked)

	content := container.Center(btTest)

	return content
}

func btTestClicked() {
	mainWindow.SetContent(SomeOtherContent)
}
```

In this exemple "mainWindow" is a global var defined in my package, so I can access it from anywhere. But I don't really like this solution.
Maybe I'm wrong to dislike it, but I'd rather share a reference to my window. But as there is no way to pass params to a button function,
I guess I need to write my button function inside the NewContent function ? Or keep a global var holding my window. 

I'm a bit new to Golang and I don't pretend to know what's good or not good habits with this language. But from my experience global context is
something I tend to avoid as much as possible. I'm probably not the only one asking my self how to structure this kind of things. 

Thanks for reading and thanks in advance for your help.

Kind regards.

**Comment by andydotxyz** [comment, marked as answer]
This really is more of a Go question than a Fyne architecture question.
We don't have extra parameters to pass into button callbacks because it would cause all callbacks to need to mirror this and it's hard to know if you'd want the button or window etc etc.

However there are many ways to work with this, you could use a wrapper function:

```go
func NewContent() *fyne.Container {
	a := app.New()
	mainWindow := a.NewWindow("Thing")
	btTest := widget.NewButton("Test", func() {
		btTestClicked(mainWindow)
	})

	content := container.NewStack(btTest)
	return content
}

func btTestClicked(win fyne.Window) {
	win.SetContent(SomeOtherContent)
}
```

Or attach your functionality to a custom type:

```go
type myApp struct {
	myWindow fyne.Window
}

func NewContent() *fyne.Container {
	a := &myApp{}
	btTest := widget.NewButton("Test", a.btTestClicked)

	content := container.NewStack(btTest)

	return content
}

func (m *myApp) btTestClicked() {
	m.myWindow.SetContent(SomeOtherContent)
}
```

Both of these are shown through the examples and longer form tutorial videos.

**Reply by halsten-dev** [reply]
Yes I agree. That's why I closed the question before any answer. Thanks for you insight tho. :)
